# What is Association Rule Mining?
Association rule minining is a rule-based machine learning method for discovering interesting relations between variables in large databases. It is intended to identify strong rules discovered in databases using some measures of interestingness. In any given transaction with a variety of items, association rules are meant to discover the rules that determine how or why certain items are connected.(source: https://en.wikipedia.org/wiki/Association_rule_learning)

- **This Algorithm is very frequently used in e-commerce domain.**
- **It helps in analysis of customer interests over a set of items.**

### Lets take some example data for this tutorial
#### The data has the records of items purcahed where each element of 'Data' refers to a single transaction

In [1]:
Data = [['Power Bank', 'Screen Guard' , 'Travel Charger'],
 ['Screen Guard', 'Bluetooth Headset', 'Mobile Cover'],
 ['Screen Guard','Arm Band','Mobile Cover'],
 ['Power Bank','Screen Guard','Leather Pouch'],
 ['Bluetooth Headset', 'Power Bank' , 'Mobile Cover']]


## We need to know some important terms before diving into Association Rule Mining(ARM)
- Support
- Confidence
- Itemsets
- Lift

### Support
Support is the number of times you see an item or items over a list of all the transactions.

#### Representing Support
Support of two items A and B is represented as:
support(A->B) = P(AUB)

The support of A -> B is the percentage of transactions that contain both A and B.

A could be Mobile Chargers
B could be Adapters
Support of A -> B = Number of Transactions that have both A and B / Total Number of Transactions

### Support Calculation
Consider the below set of transaction

- basket1 = {'vanilla wafers', 'bananas' , 'dog food'}
- basket2 = {'bananas', 'bread', 'yogurt'}
- basket3 = {'bananas','apples','yogurt'}
- basket4 = {'vanilla wafers','bananas','whipped cream'}
- basket5 = {'bread', 'vanilla wafers' , 'yogurt'}
- basket6 = {'milk' 'bread' 'bananas'}
- basket7 = {'vanilla wafers', 'apples' , 'bananas'}
- basket8 = {'yogurt', 'apples', 'vanilla wafers'}
- basket9 = {'vanilla wafers', 'bananas' , 'milk'}
- basket10 =  {'bananas', 'bread', 'peanut butter'}

If we calculate the support for each item we will get

* apples 3
* bananas 8
* bread 4
* dog food 1
* milk 2
* peanut butter 1
* vanilla wafers 6
* yogurt 4
* whipped cream 1


When the number expressed above is divided by total number of transactions, we get the percentage value.
The above representation of items(it can be one item or more) with support are known as itemsets.

### Confidence
In the previous steps , you have understood how to identify all the frequent items in your baskets.
After identifying them , the next logical step is to see if one of the items is triggering purchase of another item

Confidence is a directional relationship between two or more items.

It is represented in the following manner:

*confidence(A->B) = P(B|A)*

We can read this as the confidence of item A leading to item B is the probability of B given A.
Another way of writing confidence is

*confidence(A->B) = support(AUB) / support(A)*

The confidence of A->B can be explained as percentage of baskets containing A and B divided by the percentage of baskets that containing just A.

### Lift 
Lift is a way of quantifying the support and confidence of a set of items.

Lift (x->y) = Proportion of transactions with X and Y / (proportion of transactions with x) * (prop of transactions with y)

**Consider the following Scenario**

- 0.1% of all the transactions have milk and bread i.e support of milk -> bread is 0.1%
- 0.9% of all transactions that have milk and bread also have a banana i.e confidence of milk , bread -> banana is 0.9%
- 3% of all transactions involve banana alone

X represent Milk and Bread

Y represents Banana

*Lift (Milk , Bread) -> Banana = 0.9% / (0.1% * 3%) = 3*

What does 3 signify?

A customer is 3 times more likely to purchase a banana given that he/she buys milk and bread together.

This number is not valid if each of the items is purchased individually.

## Association Rule
Getting the support and confidence are the pre-requisites for Association Rule.
After getting the metrics, we can form the Association Rule.

*vanilla wafers -> bananas, whipped cream
[support=10%, confidence=80%]*

Interpreting the above rule , 10% of the carts or baskets have vanilla wafers , banana and whipped cream together.

80% of the customers who brought vanilla wafers also purchased bananas and whipped cream

### Components of Association Rule

Let us take the rule

item A -> item B , item C 
support 10 % , confidence 60% 

Left-hand side is called antecedent

Right-hand side is called consequent

Items A , B and C are purchased in 10% of the transactions
Among all the transactions that have item A , 60% of the transactions have items B and C.

## Apriori for Association Rule Mining

Apriori Algorithm is a very efficient mechanism for generating Association Rules.

The main objective of this algorithm is to determine all the possible rules that satisfy the required support and confidence constraints.

#### Steps - Singleton Set
- First identify all possible single item steps
- Filter those items which do not satisfy the minimum support
- Retain only those items that satisfy the support and confidence threshold
#### Steps - Doubleton Set
- Find all two item sets
- Filter all items with minimum support
- Find all possible rules from this set
- Filter only those rules with minimum confidence
- Add rule to list of rules and move on
#### Apriori - Extending the itemset
- The steps mentioned in the previous cards can be extended to multiple items and then determining all the possible rules that satisfy the constraints.
- Finally it is advised to stop when the item set cannot be made any bigger.
#### Challenges
Depending on the data collected, sometimes:
- There can be too few item sets that satisfy the minimum support
- This might lead to missing out on strong associations because of lack of support

 #### Run the below cell to import necessary packages to perform Association Rule Mining
 - In this exercise , you will try out ASM regression using mlxtend library

In [2]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.preprocessing import TransactionEncoder
import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

### Find all the Frequent Item sets
 - Initalize TransactionEncoder as `te`
 - TransactionEncoder works in the same way as pd.Dummies/sklearn.preprocessing.OneHotEncoder as in it creates one hot encoding for the different items in across transactions.
 - Fit and transform the transaction data('Data') to perform Transaction Encoding and store the result in variable `te_ary`
 - Create a dataframe of Transaction Encoded Data and store it in varaible `dataFrame`



In [3]:

te = TransactionEncoder()
te_ary = te.fit(Data).transform(Data)
dataFrame = pd.DataFrame(te_ary, columns=te.columns_)
dataFrame.head()

Arm Band  Bluetooth Headset  Leather Pouch  Mobile Cover  Power Bank  Screen Guard  Travel Charger
0     False              False          False         False        True          True            True
1     False               True          False          True       False          True           False
2      True              False          False          True       False          True           False
3     False              False           True         False        True          True           False
4     False               True          False          True        True         False           False

 - Find all frequent item sets with minimum support **0.1** using **apriori function** and store the answer in the variable `frequent_itemsets`

In [4]:
frequent_itemsets =apriori(dataFrame, min_support=0.1, use_colnames=True)
frequent_itemsets

support                                         itemsets
0       0.2                                       (Arm Band)
1       0.4                              (Bluetooth Headset)
2       0.2                                  (Leather Pouch)
3       0.6                                   (Mobile Cover)
4       0.6                                     (Power Bank)
5       0.8                                   (Screen Guard)
6       0.2                                 (Travel Charger)
7       0.2                         (Arm Band, Mobile Cover)
8       0.2                         (Arm Band, Screen Guard)
9       0.4                (Bluetooth Headset, Mobile Cover)
10      0.2                  (Bluetooth Headset, Power Bank)
11      0.2                (Bluetooth Headset, Screen Guard)
12      0.2                      (Leather Pouch, Power Bank)
13      0.2                    (Leather Pouch, Screen Guard)
14      0.2                       (Mobile Cover, Power Bank)
15      0.4                     (Mobile Cover, Screen Guard)
16      0.4                       (Screen Guard, Power Bank)
17      0.2                     (Travel Charger, Power Bank)
18      0.2                   (Screen Guard, Travel Charger)
19      0.2           (Arm Band, Mobile Cover, Screen Guard)
20      0.2    (Bluetooth Headset, Mobile Cover, Power Bank)
21      0.2  (Bluetooth Headset, Mobile Cover, Screen Guard)
22      0.2        (Leather Pouch, Screen Guard, Power Bank)
23      0.2       (Screen Guard, Travel Charger, Power Bank)

#### Generate association rules for all the frequent_itemsets
- Generating association rules for all the frequent_itemsets with minimum confidence 0.7 and storing in variable `association_rule`. 


In [5]:

association_rule = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.7)
association_rule

antecedents                   consequents  antecedent support  consequent support  support  confidence      lift  leverage  conviction  zhangs_metric
0                          (Arm Band)                (Mobile Cover)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
1                          (Arm Band)                (Screen Guard)                 0.2                 0.8      0.2         1.0  1.250000      0.04         inf       0.250000
2                 (Bluetooth Headset)                (Mobile Cover)                 0.4                 0.6      0.4         1.0  1.666667      0.16         inf       0.666667
3                     (Leather Pouch)                  (Power Bank)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
4                     (Leather Pouch)                (Screen Guard)                 0.2                 0.8      0.2         1.0  1.250000      0.04         inf       0.250000
5                    (Travel Charger)                  (Power Bank)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
6                    (Travel Charger)                (Screen Guard)                 0.2                 0.8      0.2         1.0  1.250000      0.04         inf       0.250000
7            (Arm Band, Mobile Cover)                (Screen Guard)                 0.2                 0.8      0.2         1.0  1.250000      0.04         inf       0.250000
8            (Arm Band, Screen Guard)                (Mobile Cover)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
9                          (Arm Band)  (Mobile Cover, Screen Guard)                 0.2                 0.4      0.2         1.0  2.500000      0.12         inf       0.750000
10    (Bluetooth Headset, Power Bank)                (Mobile Cover)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
11         (Mobile Cover, Power Bank)           (Bluetooth Headset)                 0.2                 0.4      0.2         1.0  2.500000      0.12         inf       0.750000
12  (Bluetooth Headset, Screen Guard)                (Mobile Cover)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
13      (Leather Pouch, Screen Guard)                  (Power Bank)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
14        (Leather Pouch, Power Bank)                (Screen Guard)                 0.2                 0.8      0.2         1.0  1.250000      0.04         inf       0.250000
15                    (Leather Pouch)    (Screen Guard, Power Bank)                 0.2                 0.4      0.2         1.0  2.500000      0.12         inf       0.750000
16     (Screen Guard, Travel Charger)                  (Power Bank)                 0.2                 0.6      0.2         1.0  1.666667      0.08         inf       0.500000
17       (Travel Charger, Power Bank)                (Screen Guard)                 0.2                 0.8      0.2         1.0  1.250000      0.04         inf       0.250000
18                   (Travel Charger)    (Screen Guard, Power Bank)                 0.2                 0.4      0.2         1.0  2.500000      0.12         inf       0.750000

## Some Inferences
### Support value for Leather Pouch -> Screen Guard is 0.2, so among all transactions Leather Pouch and Screen gaurd are bought together in 20% of them.
### The lift value for (Arm Band, Mobile Cover)->(Screen Guard) is 1.25, which implies that a person who bought Arm Band and Mobile Cover is 1.25 likely to also buy Screen Gaurd


